src/Cli/dotnet/ReleasePropertyProjectLocator.cs

```diff
@@ -0,0 +1,153 @@
+{
+    class ReleasePropertyProjectLocator : ProjectLocator
+    {
+        /// <param name="slnProjectPropertytoCheck">A property to enforce if we are looking into SLN files. If projects disagree on the property, throws exception.</param>
+        /// <returns>A project instance that will be targeted to publish/pack, etc. null if one does not exist.</returns>
+        public override ProjectInstance GetTargetedProject(IEnumerable<string> slnOrProjectArgs, string slnProjectPropertytoCheck = "")
+        {
+            string potentialProject = "";
+
+            foreach (string arg in slnOrProjectArgs.Append(Directory.GetCurrentDirectory()))
+            {
+                if (IsValidProjectFilePath(arg))
+                {
+                    return TryGetProjectInstance(arg);
+                }
+                else if (Directory.Exists(arg)) // We should get here if the user did not provide a .proj or a .sln
+                {
+                    try
+                    {
+                        return TryGetProjectInstance(MsbuildProject.GetProjectFileFromDirectory(arg).FullName);
+                    }
+                    catch (GracefulException)
+                    {
+                        string potentialSln = Directory.GetFiles(arg, "*.sln", SearchOption.TopDirectoryOnly).FirstOrDefault();
+
+                        if (!string.IsNullOrEmpty(potentialSln))
+                            return GetSlnProject(potentialSln, slnProjectPropertytoCheck);
+                    } // If nothing can be found: that's caught by MSBuild XMake::ProcessProjectSwitch -- don't change the behavior by failing here. 
+                }
+            }
+
+            return string.IsNullOrEmpty(potentialProject) ? null : TryGetProjectInstance(potentialProject);
+        }
+
+        /// <returns>The top-level project (first if multiple exist) in a SLN. Returns null if no top level project. Throws exception if two top level projects disagree
+        /// in the configuration property to check.</returns>
+        public override ProjectInstance GetSlnProject(string slnPath, string slnProjectConfigPropertytoCheck = "")
+        {
+            if (Environment.GetEnvironmentVariable("ENABLE_P_RELEASE_SLN") == null) // This has a performance overhead so don't do this unless opted in.
+                return null; // The user will be warned if they do not have this set and try this scenario with one of the properties set.
+
+            SlnFile sln;
+            try
+            {
+                sln = SlnFileFactory.CreateFromFileOrDirectory(slnPath);
+            }
+            catch (GracefulException)
+            {
+                return null; // This can be called if a solution doesn't exist. MSBuild will catch that for us.
+            }
+
+            List<ProjectInstance> configuredProjects = new List<ProjectInstance>();
+            HashSet<string> configValues = new HashSet<string>();
+            bool shouldReturnNull = false;
```

Comment:
I thought `Add` was an `atomic` operation and the order did not matter here, so didn't lock as it would decrease performance. Looks like I was wrong and `System.Collections.Concurrent` is the only thread-safe-lib. Oops, thank you!

Reply:
Also, thank goodness for RAII locks